instrumentation/opentelemetry-instrumentation-django/src/opentelemetry/instrumentation/django/middleware/sqlcommenter_middleware.py

```diff
@@ -0,0 +1,183 @@
+#!/usr/bin/python
+#
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import absolute_import
+
+import logging
+import sys
+
+import django
+from django.db import connection
+from django.db.backends.utils import CursorDebugWrapper
+
+if sys.version_info.major <= 2:
+    import urllib
+
+    url_quote_fn = urllib.quote  # pylint: disable=maybe-no-member
+else:
+    import urllib.parse
+
+    url_quote_fn = urllib.parse.quote
+
+try:
+    from opentelemetry.trace.propagation.tracecontext import (
+        TraceContextTextMapPropagator,
+    )
+
+    propagator = TraceContextTextMapPropagator()
+except ImportError:
+    propagator = None
+
+django_version = django.get_version()
+logger = logging.getLogger(__name__)
+
+KEY_VALUE_DELIMITER = ","
+
+
+class SqlCommenter:
+    """
+    Middleware to append a comment to each database query with details about
+    the framework and the execution context.
+    """
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        with connection.execute_wrapper(QueryWrapper(request)):
+            return self.get_response(request)
+
+
+class QueryWrapper:
+    def __init__(self, request):
+        self.request = request
+
+    def __call__(self, execute, sql, params, many, context):
+        # pylint: disable-msg=too-many-locals
+        _with_framework = getattr(
+            django.conf.settings, "SQLCOMMENTER_WITH_FRAMEWORK", True
+        )
+        _with_controller = getattr(
+            django.conf.settings, "SQLCOMMENTER_WITH_CONTROLLER", True
+        )
+        _with_route = getattr(
+            django.conf.settings, "SQLCOMMENTER_WITH_ROUTE", True
+        )
+        _with_app_name = getattr(
+            django.conf.settings, "SQLCOMMENTER_WITH_APP_NAME", False
+        )
+        _with_opencensus = getattr(
+            django.conf.settings, "SQLCOMMENTER_WITH_OPENCENSUS", False
+        )
+        _with_opentelemetry = getattr(
+            django.conf.settings, "SQLCOMMENTER_WITH_OPENTELEMETRY", False
```

Comment:
Context propagation should be enabled by default. That is one most important this this middleware can provide for correlation.